Task 2/script.js

```diff
@@ -1,3 +1,145 @@
+// было бы еще понимать как их использовать вместе с фронтом, потому что у меня вот нималейшей идеи. Выглядит так будто создаешь какую-то программу с нуля
+// которой никто не будет пользоваться потому что в мире все это и так уже давно создали и выложили в открытый доступ с фиксами и т.д. Но для практики, супер.
+// Ладно, как начну писать... сорри. Основные момент где хотелось бы разобраться я помечал цифрами в комментариях что--бы потом было легче ссылаться. Приступаем:
 
+function createBookId() {
+    // return '#' + Math.floor(Math.random()*10000);  (2) Если создавать ID вместе с # то потом не получается удалить книгу функцией removeBook.
+    // С чем связано понять не могу.
+    return Math.floor(Math.random() * 10000);
+}
+
+// функция конструктор для создания книги
+function Book(title, author, year) {
+    this.id = createBookId();
+    this.title = title;
+    this.author = author;
+    this.year = year;
+    this.available = true;
+    this.borrowed = false;
+    this.ratings = [];
+
+    function createBookId() {
+        // return '#' + Math.floor(Math.random()*10000);  (2) Если создавать ID вместе с # то потом не получается удалить книгу функцией removeBook.
+        // С чем связано понять не могу.
+        return Math.floor(Math.random() * 10000);
+    }
+
+    // добавляем рейтинг, но не даем добавить если пользователь книгу не брал.
+    this.addRating = function (user, rating) {
+        if (this.borrowed && user.borrowedBooks.includes(this.id)) {
+            this.ratings.push(rating);
+        } else {
+            console.log("This user hasn't borrowed this book.");
+        }
+    }
+    // полностью подсмотрел на форуме по подсчету среднего рейтинга. Короче как ты понял я тут многое не писал из головы, чтото из старых дз либо из инета)))
+    // в целом то логично, просто мне трудно было догадаться что .length отвечает за суммарное количество оценок)) Ниче, качаемся как говорится
+    this.getAverageRating = function () {
+        if (this.ratings.length === 0) return 0;
+        const sum = this.ratings.reduce((acc, curr) => acc + curr, 0);
+        return sum / this.ratings.length;
+    }
+}
+
+// Конструктор по созданию библиотеки. 
+function Library() {
+    this.books = [];
+
+    this.addBook = function (book) {
+        this.books.push(book);
+    }
+    // вначале сделал как в кофейне, типа через findIndex. А потом гдето в видео увидел что оказывается можно переписать просто массив, добавив все, кроме заданного.
+    // все гениальное просто)))
+    this.removeBook = function (bookId) {
+        this.books = this.books.filter(book => book.id !== bookId);
+        return console.log(`You've removed book with id ${bookId}`)
+    }
+
+    this.findBooksByAuthor = function (author) {
+        return this.books.filter(book => book.author === author);
+    }
+
+    this.listAvailableBooks = function () {
+        return this.books.filter(book => book.available === true);
```

Comment:
```suggestion
        return this.books.filter(book => book.available);
```

оно ж уже возвращает тебе true либо false, так что сравнение redundant